### DISCUSSION TITLE
Default remote push url in Magit

### DISCUSSION THREAD

**Discussion by josephmturner** [magit/magit#5249] [Support] [answered]
A common workflow:

1. Clone a repo, e.g., `https://github.com/magit/transient/`, and say
"no" when asked to set it as the default push remote.

2. "Fork" it on Github, e.g., to `https://github.com/josephmturner/transient/`

3. Inform Magit of this new remote.  `magit-remote-add` prompts me for a
new name, so I type in `josephmturner` and it correctly guesses my
remote's `url`: `https://github.com/josephmturner/transient/`

4. Inform Magit of the `pushurl`.  For this, I need to type in the full
`pushurl`: `git@github.com:josephmturner/transient/` or use future history
and then edit the https link into a ssh filepath.

Is it possible to automate step 4?

Perhaps Magit could automatically assign a remote's `pushurl` iff its `http`
`url` matches one of the regexps in a customizable list, e.g.,

```
("^https:\/\/github\.com\/josephmturner\/.*"
 "^https:\/\/git\.sr\.ht\/breatheoutbreathein\/.*")
```

Then maybe another `defcustom` could control converting a remote `url` to a
`pushurl`.  Here's a *rough* draft of what might work for me:

```
(defun magit-convert-url-to-pushurl (url)
  (pcase-let* (((cl-struct url host filename) (url-generic-parse-url url))
               (`(,username ,project) (split-string filename "/" t)))
    (format "git@%s:%s/%s" host username project)))

(magit-convert-url-to-pushurl "https://github.com/josephmturner/transient/")
  =>  "git@github.com:josephmturner/transient"
```

I use ssh urls for pushing but not pulling because I don't want to type
in my ssh password just to fetch the latest changes.

Thank you!!

Joseph

**Comment by kyleam** [comment, marked as answer]
[ related to gh-5004 ]

> Is it possible to automate step 4?

One way would be to configure that at the Git level.  From `git-config(1)`:

```
url.<base>.pushInsteadOf
    Any URL that starts with this value will not be pushed to; instead, it will be rewritten
    to start with <base>, and the resulting URL will be pushed to. In cases where some site
    serves a large number of repositories, and serves them with multiple access methods, some
    of which do not allow push, this feature allows people to specify a pull-only URL and have
    Git automatically use an appropriate URL to push, even for a never-before-seen repository
    on the site. When more than one pushInsteadOf strings match a given URL, the longest match
    is used. If a remote has an explicit pushurl, Git will ignore this setting for that
    remote.
```

> I use ssh urls for pushing but not pulling because I don't want to type
> in my ssh password just to fetch the latest changes.

I suspect most SSH users avoid repeatedly entering their password by setting up ssh-agent.

**Reply by josephmturner** [reply]
Thank you, Kyle!!  `pushInsteadOf` meets my need perfectly.  I've added:

```
[url "git@git.sr.ht:"]
  pushInsteadOf = "https://git.sr.ht/"
[url "git@codeberg.org:"]
  pushInsteadOf = "https://codeberg.org/"
[url "git@github.com:"]
  pushInsteadOf = "https://github.com/"
```
to my global git config.

> I suspect most SSH users avoid repeatedly entering their password by setting up ssh-agent.

I agree it's convenient, but... being forced to enter my password has saved me from quickly pushing when I needed to think twice instead. :)

Thanks for your help!!

Joseph
